int32 row
int32 column
int32 angle